### PR TITLE
Drop datamodel-code-generator artifacts in favor of pure jsonschema artifacts + conversions

### LINKED ISSUE #125: Remove datamodel-code-generator artifacts, and replace `_convert` logic with jsonschema manipulation only
- [ ] Remove `datamodel-code-generator` dependency from `wt-compiler`
- [ ] Remove formdata.py and params.py from `WorkflowArtifacts`
- [ ] Remove generation of formdata.py and params.py from `compiler.py` logic
- [ ] In `wt-compiler/src/wt_compiler/templates/**`, update any behavior that relies on Formdata and/or Params models to use simple JSON (i.e. dict) manipulation and/or https://github.com/python-jsonschema/jsonschema for validation (i.e. 422 checks)

### PR DESCRIPTION
**Pull request by cisaacstern**
## :earth_americas: Summary

closes #125

## :package: Proposed Changes

- Implementation of #125

### CODE REVIEW

File: wt-compiler/src/wt_compiler/templates/pkg/dags/run_sequential.jinja2 [hunk 1]
```diff
@@ -28,17 +31,16 @@ from wt_task import task
...
-
 
-def main(params: Params):
+def main(params: dict[str, Any], validate_params_schema: bool = True):
     {%- if testing %}
     warnings.warn("This test script should not be used in production!")  # 🧪
     {% endif %}
-    params_dict = json.loads(params.model_dump_json(exclude_unset=True))
+    if validate_params_schema:
+        _validate(params, _metadata.load_params_schema())
```
Comment (cisaacstern) [inline wt-compiler/src/wt_compiler/templates/pkg/dags/run_sequential.jinja2:40]: I chose to put `validate_params_schema` (defaulting to `True`) here so that direct callers of `main()` (testing scenarios, or just because of Hyrum's Law) don't get into some weird error state due to malformed `params` dict. That being said, when called from the cli -> dispatch() -> main() we will pass this as `False` because we will have already validated the params at the cli layer and don't need to repeat that again here.

File: wt-compiler/src/wt_compiler/templates/pkg/cli.jinja2 [hunk 2]
```diff
@@ -6,11 +6,13 @@ import sys
...
 from io import TextIOWrapper
 from pathlib import Path
-from typing import Optional
+from typing import Any, Optional
 from urllib.parse import urlparse
 
 import click
 
+from .metadata import load_params_schema, load_rjsf_schema
+
```
Comment (cisaacstern) [inline wt-compiler/src/wt_compiler/templates/pkg/cli.jinja2]: hmm i think the convention in this template is to defer imports into the bodies of the handler functions, so we may want to move these

File: wt-compiler/src/wt_compiler/templates/pkg/dispatch.jinja2 [hunk 3]
```diff
@@ -24,7 +25,7 @@ def dispatch(
             raise ValueError(f"Invalid execution mode: {execution_mode}")
 
     try:
-        result = dispatcher(params=params)
+        result = dispatcher(params=params, validate_params_schema=validate_params_schema)
```
Comment (cisaacstern) [inline wt-compiler/src/wt_compiler/templates/pkg/dispatch.jinja2:28]: defaulting to `validate_params_schema=True` here for reasons noted above, but the most common call path from cli.py will pre-validate params and set this to `False`

File: wt-compiler/tests/test_wizard_default.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 from pathlib import Path
 
 import pytest
-import yaml
+import ruamel.yaml
```
Comment (cisaacstern) [inline wt-compiler/tests/test_wizard_default.py:10]: `datamodel-code-generator` was pulling in PyYAML transitively, and without it we no longer have PyYAML, so changed this to get tests passing.

File: wt-contracts/src/wt_contracts/formdata.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+The conversion is purely structural: top-level keys in ``params_schema``
+identify direct (flat) tasks, while any other top-level key in ``rjsf_schema``
+is treated as a task group whose nested ``properties`` enumerate its tasks.
+The helpers do not read RJSF custom keywords (e.g. ``ecoscope:task_group``);
+group membership is inferred from schema structure alone.
+"""
+
+from typing import Any
+
+import jsonschema  # type: ignore[import-untyped]
```
Comment (cisaacstern) [inline wt-contracts/src/wt_contracts/formdata.py]:
```suggestion
import jsonschema
```

let's remove this ignore and use `types-jsonschema` from typeshed to get mypy coverage on this import.

File: wt-contracts/src/wt_contracts/formdata.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    """
+    validate(formdata, rjsf_schema)
+    flat_keys = set(params_schema.get("properties", {}).keys())
+    out: dict[str, Any] = {}
+    for k, v in formdata.items():
+        if k in flat_keys:
+            out[k] = v
+        else:
+            for inner_k, inner_v in v.items():
+                out[inner_k] = inner_v
```
Comment (cisaacstern) [inline wt-contracts/src/wt_contracts/formdata.py]: Is this actually correct? I would have thought this would need to be recursive? And also, where is the check for `flat_keys` membership here?

File: wt-contracts/src/wt_contracts/formdata.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    """Raised when a dict fails JSON schema validation.
+
+    Attributes:
+        errors: List of serialized ``jsonschema`` errors. Each entry has
+            ``message``, ``path``, ``schema_path``, and ``validator`` keys.
+    """
+
+    def __init__(self, errors: list[dict[str, Any]]):
+        self.errors = errors
+        super().__init__(f"{len(errors)} validation error(s): {errors}")
```
[the remark this answers is not in the document]
Comment (cisaacstern) [reply wt-contracts/src/wt_contracts/formdata.py:94]: i'll take my chance on a noisy error, i'd rather have the information i need

File: wt-runner/src/wt_runner/app.py [hunk 8]
```diff
@@ -656,13 +656,51 @@ async def data_connection_property_names(
...
+
+    Returns:
+        List of FastAPI-shaped error dicts.
+    """
+    return [
+        {
+            "loc": list(e.get("path", [])),
+            "msg": e.get("message", ""),
+            "type": e.get("validator", "value_error"),
+        }
```
[the remark this answers is not in the document]
Comment (cisaacstern) [reply wt-runner/src/wt_runner/app.py]: this method was removed in the latest commits, we are no longer trying to match a fastapi 422 response shape

File: wt-contracts/src/wt_contracts/formdata.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+    Raises:
+        ValidationError: If ``instance`` does not validate against ``schema``.
+            The exception's ``errors`` attribute contains a list of serialized
+            validation errors with stable, JSON-serializable shape.
+    """
+    validator = jsonschema.Draft202012Validator(schema)
+    errors = sorted(validator.iter_errors(instance), key=lambda e: list(e.absolute_path))
+    if errors:
+        raise ValidationError(_serialize_errors(errors))
```
[the remark this answers is not in the document]
Comment (cisaacstern) [reply wt-contracts/src/wt_contracts/formdata.py]: good point, fixed in [b0fa60d](https://github.com/wildlife-dynamics/wt/pull/150/commits/b0fa60dba0f4c492be2b5d175a545d3e05062438)

File: wt-contracts/src/wt_contracts/formdata.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        Validating ``{"age": "old"}`` against a schema that requires
+        ``{"age": {"type": "integer"}}`` produces a single error of the form::
+
+            {
+                "message": "'old' is not of type 'integer'",
+                "path": ["age"],
+                "schema_path": ["properties", "age", "type"],
+                "validator": "type",
+                "input": "old",
+            }
```
Comment (cisaacstern) [inline wt-contracts/src/wt_contracts/formdata.py]: this is not a real doctest. please wrap in backticks ```python and make this a true doctest

File: wt-contracts/tests/test_formdata.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
+"""Tests for wt_contracts.formdata."""
```
Comment (cisaacstern) [inline wt-contracts/tests/test_formdata.py:1]: how comprehensive are these tests of the conversion functions? what is the actual logical branch and/or line coverage of the the functions in formadata.py? we want to be sure we're getting this right. we don't want tests for tests sake but if there are edge or corner cases or just regular cases we're not testing, we should test them

File: wt-runner/src/wt_runner/app.py [hunk 12]
```diff
@@ -656,13 +657,29 @@ async def data_connection_property_names(
...
+class _ConvertValidationError(Exception):
+    """Raised when the compiled CLI's ``convert`` returns ``validation_errors``.
+
+    Attributes:
+        errors: List of jsonschema-native error dicts emitted by the CLI.
+    """
+
+    def __init__(self, errors: list[dict[str, Any]]):
+        self.errors = errors
+        super().__init__(f"{len(errors)} validation error(s)")
```
Comment (cisaacstern) [inline wt-runner/src/wt_runner/app.py]: what is the advantage of defining this here versus using the error model directly from wt-contracts?

File: wt-contracts/src/wt_contracts/formdata.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+        >>> try:
+        ...     validate({"age": "old"}, {"properties": {"age": {"type": "integer"}}})
+        ... except ValidationError as e:
+        ...     err = e.errors[0]
+        >>> err["message"]
+        "'old' is not of type 'integer'"
+        >>> err["path"], err["schema_path"]
+        (['age'], ['properties', 'age', 'type'])
+        >>> err["validator"], err["input"]
+        ('type', 'old')
```
Comment (cisaacstern) [inline wt-contracts/src/wt_contracts/formdata.py]: this doctest does not actually exercise the class on which the docstring is defined

File: wt-contracts/src/wt_contracts/formdata.py (hunk 6 above)
Comment (cisaacstern) [reply wt-contracts/src/wt_contracts/formdata.py]: resolved by comment on lines 192-193

File: wt-contracts/tests/test_formdata.py (hunk 11 above)
Comment (cisaacstern) [reply wt-contracts/tests/test_formdata.py:1]: coverage bumped to 100% in latest commit

File: wt-contracts/src/wt_contracts/formdata.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+
+def _serialize_errors(
+    errors: list[jsonschema.ValidationError],
+) -> list[ValidationErrorItemDict]:
+    return [
+        {
+            "message": e.message,
+            "path": list(e.absolute_path),
+            "schema_path": list(e.absolute_schema_path),
+            "validator": e.validator if isinstance(e.validator, str) else "",
```
Comment (cisaacstern) [inline wt-contracts/src/wt_contracts/formdata.py]: what other type can e.validator be aside from str? are we loosing information here?

File: wt-contracts/src/wt_contracts/formdata.py (hunk 10 above)
Comment (cisaacstern) [reply wt-contracts/src/wt_contracts/formdata.py]: addressed in recent commits

File: wt-contracts/src/wt_contracts/formdata.py (hunk 5 above)
Comment (cisaacstern) [reply wt-contracts/src/wt_contracts/formdata.py]: addressed in recent commits

File: wt-runner/src/wt_runner/app.py (hunk 12 above)
Comment (cisaacstern) [reply wt-runner/src/wt_runner/app.py]: addressed in recent commits

File: wt-compiler/src/wt_compiler/templates/pkg/cli.jinja2 (hunk 2 above)
Comment (cisaacstern) [reply wt-compiler/src/wt_compiler/templates/pkg/cli.jinja2]: fixed in recent commits

File: wt-contracts/src/wt_contracts/formdata.py (hunk 14 above)
Comment (cisaacstern) [reply wt-contracts/src/wt_contracts/formdata.py]: addressed in latest commits

File: wt-contracts/src/wt_contracts/formdata.py (hunk 13 above)
Comment (cisaacstern) [reply wt-contracts/src/wt_contracts/formdata.py]: addressed in latest commits

**Comment by cisaacstern** [comment]
Curious if the integration tests will pass without the following patch I needed to apply locally (post-compile) to get the metadata features of the cli to work ...

```diff

--- a/ecoscope-workflows-events-workflow/pixi.toml
+++ b/ecoscope-workflows-events-workflow/pixi.toml
@@ -22,6 +22,9 @@ linux = "4.4.0"
 version = "==2.11.3"
 channel = "https://repo.prefix.dev/ecoscope-workflows/"

+[pypi-dependencies.wt-contracts]
+path = "../../wt/wt-contracts"
+
 [dependencies.click]
 version = ">=8.0.0,<9.0.0"
 channel = "https://conda.anaconda.org/conda-forge/"
 
```

File: wt-contracts/src/wt_contracts/formdata.py [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+    out: dict[str, Any] = {}
+    for k, v in params.items():
+        if k in rjsf_props and k not in task_groups:
+            out[k] = v
+            continue
+        if k in rjsf_props and k in task_groups:
+            # k is itself a group name in rjsf -- shouldn't happen for flat params,
+            # but if so, treat it as direct.
+            out[k] = v
+            continue
```
Comment (atmorling) [inline wt-contracts/src/wt_contracts/formdata.py:255]: Interested to know if this is the same:

```suggestion
        if k in rjsf_props:
            out[k] = v
            continue
```

File: wt-compiler/src/wt_compiler/jsonschema.py [hunk 16]
```diff
@@ -267,3 +267,30 @@ def apply_overrides(
...
+        silently mis-target. ``$defs`` is shared between the two artifacts,
+        so its overrides apply cleanly to both.
+
+        Args:
+            initial: The initial configuration to override
+
+        Returns:
+            A new configuration with only ``$defs`` overrides applied
+        """
+        initial_defs = copy.deepcopy(initial.definitions)
```
Comment (cisaacstern) [inline wt-compiler/src/wt_compiler/jsonschema.py:289]: Per sync w/ @atmorling:

- This a fix for the the follow diff being required to get params.json emitted by Ecoscope Platform standard workflows (that use ValueGrouper) to be valid jsonschema: https://github.com/ecoscope-platform-workflows-releases/events/commit/3955751925f1edea8266ead20f216b447599b070#diff-2d2650b30ffa08ea1e74e0c6ad197aa3f000068c5b8f0160225f90b715bc3cd7

Two follow-ons we noted:

1. Change ecoscope-platform package ValueGrouper json schema so that this workaround is not strictly required (though this workaround is useful regardless, and does not need to be reverted)
2. Add a pre-artifact-dump check in wt-compiler to ensure that params.json and rjsf.json are both actually valid json schema, with a descriptive error if they are not

**Review by atmorling — APPROVED** [review]
